File: Caesars-Cipher/caesar-cipher.js
/*
One of the simplest and most widely known ciphers is a Caesar cipher, also known as a shift cipher. In a shift cipher the meanings of the letters are shifted by some set amount.

A common modern use is the ROT13 cipher, where the values of the letters are shifted by 13 places. Thus 'A' ↔ 'N', 'B' ↔ 'O' and so on.

Write a function which takes a ROT13 encoded string as input and returns a decoded string.

All letters will be uppercase. Do not transform any non-alphabetic character (i.e. spaces, punctuation), but do pass them on.
*/

// Solution 1
function rot13(str) {
  // Split str into a character array
  return (
    str
      .split("")
      // Iterate over each character in the array
      .map.call(str, function(char) {
        // Convert char to a character code
        var x = char.charCodeAt(0);
        // Checks if character lies between A-Z
        if (x < 65 || x > 90) {
          return String.fromCharCode(x); // Return un-converted character
        }
        //N = ASCII 78, if the character code is less than 78, shift forward 13 places
        else if (x < 78) {
          return String.fromCharCode(x + 13);
        }
        // Otherwise shift the character 13 places backward
        return String.fromCharCode(x - 13);
      })
      .join("")
  ); // Rejoin the array into a string
}

/*
Code Explanation
* A string variable nstr is declared and initialized to store the decoded string.
* The for loop is used to loop through each character of the input string.
* If the character is not uppercase English alphabets(i.e. its ascii doesn’t lie between 65 and 91 ), we’ll leave it as it is and continue 78 with next iteration.
* If it’s the uppercase English alphabet, we’ll subtract 13 from its ascii code.
* If the ascii code is less than 78, it’ll get out of range when subtracted by 13 so we’ll add 26 (number of letters in English alphabets) to it so that after A it’ll go back to Z. e.g. M(77) <==> 77-13 = 64(Not an English alphabet) +26 = 90 <==> Z(90).
*/

// Solution 2
// Solution with Regular expression and Array of ASCII character codes
function rot13(str) {
  var rotCharArray = [];
  var regEx = /[A-Z]/;
  str = str.split("");
  for (var x in str) {
    if (regEx.test(str[x])) {
      // A more general approach
      // possible because of modular arithmetic
      // and cyclic nature of rot13 transform
      rotCharArray.push(((str[x].charCodeAt() - 65 + 13) % 26) + 65);
    } else {
      rotCharArray.push(str[x].charCodeAt());
    }
  }
  str = String.fromCharCode.apply(String, rotCharArray);
  return str;
}
/*
Code Explanation
* A string variable nstr is declared and initialized to store the decoded string.
* The for loop is used to loop through each character of the input string.
* If the character is not uppercase English alphabets(i.e. its ascii doesn’t lie between 65 and 91 ), we’ll leave it as it is and continue 78 with next iteration.
* If it’s the uppercase English alphabet, we’ll subtract 13 from its ascii code.
* If the ascii code is less than 78, it’ll get out of range when subtracted by 13 so we’ll add 26 (number of letters in English alphabets) to it so that after A it’ll go back to Z. e.g. M(77) <==> 77-13 = 64(Not an English alphabet) +26 = 90 <==> Z(90).
*/

File: Caesars-Cipher/caesar-cipher.cpp
#include <iostream>
#include <string>

using namespace std;

void caesarShift(string& phrase, int shiftAmount);

int main() {

    string phrase;
    int shiftAmount;

    // Get the shift parameters

    cout << "Enter your phrase: ";
    getline(cin, phrase);

    cout << "Enter your shift amount: ";
    cin >> shiftAmount;

    caesarShift(phrase, shiftAmount);

    cout << endl << phrase;

    return 0;
}

void caesarShift(string& phrase, int shiftAmount) {
    // Caesar Shift a certain phrase by a certain amount

    int normalizedAmount = shiftAmount % 26;  // Ensure that the shiftAmount keeps letters between a-z

    for (int i = 0; i < phrase.length(); i++) {  // Shift each letter

        if (phrase[i] >= 65 && phrase[i] <= 90 && phrase[i] + normalizedAmount > 90)  // Uppercase overflow
            phrase[i] = phrase[i] + normalizedAmount - 26;
        
        else if (phrase[i] >= 97 && phrase[i] <= 122 && phrase[i] + normalizedAmount > 122)  // Lowercase overflow
            phrase[i] = phrase[i] + normalizedAmount - 26;

        else if (!(phrase[i] >= 65 && phrase[i] <= 90) && !(phrase[i] >= 97 && phrase[i] <= 122))  // Leave all non-letter characters as they are
            continue;

        else
            phrase[i] = phrase[i] + normalizedAmount;
    }
}